## Calculando a similaridade entre vetores e dados

### Exercicio 1: Cálculo de distância euclidiana

Descrição: escreva uma função em python que receba dois vetores (lista de números) de mesma dimensão e retorna a distância euclidiana entre eles.

In [23]:
def calc_distancia_euclidiana(vetor1, vetor2):
    """
    Calcula a distância euclidiana entre dois vetores.
    :param vetor1: lista de números representando o primeiro vetor
    :param vetor2: lista de números representando o segundo vetor
    :return: A distancia euclidiana entre os dois vetores
    """
    
    
    if len(vetor1) != len(vetor2):
        raise ValueError("Os vetores devem ter a mesma dimensão.")
    
    soma_quadrados = sum((v1 - v2) ** 2 for v1, v2 in zip(vetor1, vetor2))
    distancia = soma_quadrados ** 0.5 #retorna a raiz quadrada da soma
    return distancia

vetor_a = [1, 2, 3]
vetor_b = [4, 5, 6]

distancia = calc_distancia_euclidiana(vetor_a, vetor_b)
print(f'distancia euclidian: {distancia}')

distancia euclidian: 5.196152422706632


O valor 5.196... representa a distancia euclidiana entre os dois vetores fornecidos. A distancia euclidiana, também conhecida como norma L2 ou norma euclidiana, é uma medida de espaço euclidiano, que é o espaço geométrico mais comum.

Para interprestar esse valor:
- **Contexto geométrico:** Imagine que cada vetor representa um ponto em um espaço tridimensional (pois cada vetor tem três componentes). O valor 5.196 é a distância "em linha reta" entre esses dois pontos. Se você pudesse traçar uma linha entre os pontos representados pelos vetores [1, 2, 3] e [4, 5, 6], essa linha teria comprimento de aproximadamente 5.196 unidades.

- **Similaridade:** Em algumas aplicações, como aprendizado de máquina e estatística, a distância euclidiana pode ser usada para avaliar a similaridade entre dois conjuntos de dados. Menores distância indicam maior similaridade, enquanto maiores distância indicam menor similaridade.

- **Interpretação numérida:** O número 5.196, em si, é uma medida absoluta e sua interprestação pode depender do contexto. Por exemplo, se os vetores representam coordenadas em metros, então a distância entre os pontos é de 5.196 metros. Se não houver uma unidade específica ou um contexto, este valor é apenas uma medida de "quão longe" os pontos estão um do outro em termos matemáticos.

### Exercício 2: Distância euclidiana com validação de dados

Descrição: Modifique a função do Exercício 1 para validar se os dois vetores têm o mesmo tamanho. Se não tiverem, a função deve retornar um erro.

In [24]:
def calc_distancia_euclidiana_validacao(vetor1, vetor2):
    """
    Calcula a distância euclidiana entre dois vetores.
    :param vetor1: lista de números representando o primeiro vetor
    :param vetor2: lista de números representando o segundo vetor
    :return: A distancia euclidiana entre os dois vetores
    """
    
    
    if len(vetor1) != len(vetor2):
        raise ValueError("Os vetores devem ter a mesma dimensão.")
    
    soma_quadrados = sum((v1 - v2) ** 2 for v1, v2 in zip(vetor1, vetor2))
    distancia = soma_quadrados ** 0.5 #retorna a raiz quadrada da soma
    return distancia

# vetores com tamanhos iguais
# vetor_a = [1, 2, 3]
# vetor_b = [4, 5, 6]

# vetores com tamanhos diferentes
vetor_a = [1, 2]
vetor_b = [4, 5, 6]

try:
    distancia = calc_distancia_euclidiana_validacao(vetor_a, vetor_b)
    print(f'distancia euclidian: {distancia}')

except ValueError as e:
    print(f"Erro: {e}")


Erro: Os vetores devem ter a mesma dimensão.


### Exercítio 3: Distância euclidiana em N-Dimensões

Descrição: Generalize a função do exercício 1 para que ela possa calcular a distancia euclidiana entre dois vetores em um espaço de qualquer número de dimensões

In [25]:
def calc_distancia_euclidiana_nd(vetor1, vetor2):
    """
    Calcula a distância euclidiana entre dois vetores em um espaço qualquer número de dimensões
    
    :param vetor1: lista de números representando o primeiro vetor
    :param vetor2: lista de números representando o segundo vetor
    :return: A distancia euclidiana entre os dois vetores
    :raises: ValueError se os vetores têm amanhos diferentes 
    """
    
    
    if len(vetor1) != len(vetor2):
        raise ValueError("Os vetores devem ter a mesma dimensão.")
    
    soma_quadrados = sum((v1 - v2) ** 2 for v1, v2 in zip(vetor1, vetor2))
    distancia = soma_quadrados ** 0.5 #retorna a raiz quadrada da soma
    return distancia

vetor_a = [1, 2, 3, 4]
vetor_b = [5, 6, 7, 8]

distancia = calc_distancia_euclidiana_nd(vetor_a, vetor_b)
print(f'distancia euclidiana: {distancia}')

distancia euclidiana: 8.0


In [26]:
# usando a função
vetor_a = [0.1, 200, 3/2, 4.9, 10]
vetor_b = [1.2, 1/8, 0.78, 0.003, 20]

try:
    distancia = calc_distancia_euclidiana_nd(vetor_a, vetor_b)
    print(f'distancia euclidiana: {distancia}')
except ValueError as e:
    print(f'Error: {e}')

distancia euclidiana: 200.18922207251816


### Exercício 4: Comparando distâncias

Descrição: Escreva um script que receba três vetores da mesma dimensão. Calcule a distância euclidiana entre o primeiro e os outros dois vetores.

Imprima qual dos dois vetore está mais próximo do primeiro, baseado na distância euclidiana.

In [27]:
import numpy as np

""" Calcula a distância euclidiana entre dois vetores

:param vetor1: Lista de números representando o primeiro vetor
:param vetor2: Lista de números representando o segundo vetor
:return: A distância euclidiana entre os dois vetores
"""

def calcular_distancia(vetor1, vetor2):
    return np.linalg.norm(np.array(vetor1) - np.array(vetor2))

def vetor_mais_proximo(vetor1, vetor2, vetor3):
    """Compara as distâncias euclidianas entre o primeiro vetor e outros dois vetores.

    :param vetor1: Lista de números representando o primeiro vetor
    :param vetor2: Lista de números representando o segundo vetor
    :param vetor2: Lista de números representando o terceiro vetor
    :return: Imprime qual dos dois vetores (vetor2 ou vetor3) está mais próximo do vetor 1.
    """
    
    # Cálcula as distâncias
    distancia_1 = calcular_distancia(vetor1, vetor2)
    distancia_2 = calcular_distancia(vetor1, vetor3)
    
    # Compara as distâncias e imprime o resultado
    # Menores distâncias indicam maior similaridade, enquanto maiores distâncias indicam menor similaridade. 
    if distancia_1 < distancia_2:
        return "O vetor 2 está mais próximo do vetor 1."
    elif distancia_2 < distancia_1:
        return "O vetor 3 está mais próximo do vetor 1."
    else:
        "Os dois vetores estão à mesma distância do primeiro."

# Exemplo de uso
v1 = [1, 2, 3]
v2 = [5.0, 0.7, 1/6]
v3 = [1.4, 0.33, 2.2]

resultado = vetor_mais_proximo(v1, v2, v3)
print(f'Resultado: {resultado}')

Resultado: O vetor 3 está mais próximo do vetor 1.


### Exercício 5: Aplicação em Data Science

Descrição: Data uma matriz de dados (lista de listas, onde cada sublista é um vetor de características) e um vetor de referência, escreva uma função calculle a distância euclidiana de cada vetor na matriz até o vetor de referencia. A função deve retornar uma lista de distâncias.

In [28]:
def calcular_distancias(matriz, vetor_referencia):
    """Calcula a distância euclidiana de cada vetor em uma matriz até um vetor de referência.

    Args:
        :param matriz: Lista de listas, onde cada sublista é um vetor de características
        :param vetor_referencia: Lista de números representando o vetor de referência
        :return: Lista de distâncias euclidianas de cada vetor na matriz até o vetor de refrência.
    """
    
    distancias = []
    
    for vetor in matriz:
        distancia = calc_distancia_euclidiana(vetor, vetor_referencia)
        distancias.append(distancia)
        
    return distancias

# Matriz de exemplo
matriz_dados = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]

valor_ref = [1.2, 3.4, 0.998]

# Executa a função
distancias = calcular_distancias(matriz_dados, valor_ref)
print(f"Resultado: {distancias}")

Resultado: [2.4511230079292226, 5.9514707425980005, 10.901009311068403]
